bigchaindb/crypto/buffer.py

```diff
@@ -0,0 +1,437 @@
+# we don't use sys.maxint (= 2 ** 63 - 1) as this spec is inline with the ILP JavaScript reference implementation
+# see https://interledger.org/
+MAX_SAFE_INTEGER_JS = 2 ** 53 - 1
+
+
+class UnsignedLEB128:
+    """
+    Adapter for DWARF unsigned LEB128.
+
+    A VARUINT is a variable length integer encoded as base128 where the highest
+    bit indicates that another byte is following. The first byte contains the
+    seven least significant bits of the number represented.
+
+    see: https://en.wikipedia.org/wiki/LEB128 (ULEB128)
+    see: http://grokbase.com/t/python/python-list/112e5jpc16/encoding
+
+    """
+
+    @staticmethod
+    def encode(obj):
+        out = []
+        value = int(obj)
+        while value > INT:
+            out.append((value & REST) | MSB)
+            value /= 7
+
+        while value & MSBALL:
+            out.append((value & REST) | MSB)
+            value >>= 7
+
+        out.append(value | 0)
+        return out
+
+    @staticmethod
+    def decode(obj):
+        value = 0
+        for b in reversed(obj):
+            value = value * 128 + (ord(b) & 0x7F)
+        return value
+
+
+class Writer:
+    def __init__(self):
+        self.components = []
+
+    def write_var_uint(self, value):
+        """
+        Write a VARUINT to the stream.
+
+        Args:
+            value (int): Integer to represent.
+        """
+        out = UnsignedLEB128.encode(value)
+        self.write(out)
+
+    def write_var_bytes(self, buffer):
+        """
+        Write a VARBYTES.
+
+        A VARBYTES field consists of a VARUINT followed by that many bytes.
+
+        Args:
+            buffer (Buffer): Contents of the VARBYTES.
+        """
+        self.write_var_uint(len(buffer))
+        self.write(buffer)
+
+    def write(self, in_bytes):
+        """
+        Write a series of raw bytes.
+
+        Adds the given bytes to the output buffer.
+
+        Args:
+            in_bytes (Buffer): Bytes to write.
+        """
+        out = in_bytes
+        if isinstance(out, (list, bytearray)):
+            out = binascii.unhexlify(''.join('{:02x}'.format(x) for x in out))
+        if not isinstance(out, bytes):
+            out = out.encode('utf-8')
+        self.components.append(out)
+
+
+class Hasher(Writer):
+
+    def __init__(self, algorithm):
+        if algorithm == 'sha256':
+            self.hash = hashlib.sha256()
+        else:
+            raise NotImplementedError
+        super().__init__()
+
+    def write(self, in_bytes):
+        """
+        Adds bytes to the hash input.
+        
+        The hasher will pass these bytes into the hashing function. By overriding
+        the Writer class and implementing this method, the Hasher supports any of
+        the datatypes that a Writer can write.
+        
+        Args:
+            in_bytes (Buffer): Bytes to add to the hash.
+        """
+        out = in_bytes
+        if isinstance(out, (list, bytearray)):
```

Comment:
the write method supports chunks of bytes in a list `['bytes_1',... , 'bytes_n']`